# Explore data

First, let's see how many columns and the meaning of each columns in the datasets.

In [34]:
import pandas as pd
import numpy as np

In [35]:
# Read data from file
ds_survey_df = pd.read_csv('../Dataset/kaggle_survey_2022_responses.csv', )

# Delete first row as it is the sub-text for the meaning of the column
ds_survey_df = ds_survey_df.drop(0)

C:\Users\nk\AppData\Local\Temp\ipykernel_9416\1085557566.py:2: DtypeWarning: Columns (0,15,43,57,73,88,104,118,126,132,170,200,208,215,225,248,255,257,260,270,271,272,277,281,294) have mixed types. Specify dtype option on import or set low_memory=False.
  ds_survey_df = pd.read_csv('../Dataset/kaggle_survey_2022_responses.csv', )


In [36]:
print("Number of columns: ", len(ds_survey_df.columns))
print("Number of rows: ", len(ds_survey_df))
print("Columns: ", ds_survey_df.columns)

Number of columns:  296
Number of rows:  23997
Columns:  Index(['Duration (in seconds)', 'Q2', 'Q3', 'Q4', 'Q5', 'Q6_1', 'Q6_2', 'Q6_3',
       'Q6_4', 'Q6_5',
       ...
       'Q44_3', 'Q44_4', 'Q44_5', 'Q44_6', 'Q44_7', 'Q44_8', 'Q44_9', 'Q44_10',
       'Q44_11', 'Q44_12'],
      dtype='object', length=296)


*Notice that all columns has data types of object. This makes sense, since all the questions answer are of type string. As for Duration, the suitable data type should be float.*

---
The details meaning of the columns are aviailable as Kaggle had published them along with the survey information. 

*Note: "Person who took the survey" wiil be shortened to PWTS*

## Details meaning of each questions/columns sorted by theme:

- Durations: The duration the survey takes to complete by a person. (Calculated by system)

### Bibography:

- Q2: Age of PWTS (One choice)
- Q3: Gender of PWTS (One choice)
- Q4: Country that PWTS comes from (Once choice)
- Q5: Whether PWTS a student (One choice)

### Education:

- Q6: Platforms that PWTS have used to study Data Sciecne (Multiple choice)
- Q7: Platforms/Prodects that PWTS found to be most helpful when started studying DS (Multiple choice)
- Q8: Highest level of education that PWTS had attained/plan to attain in the next 2 years (One choice)
- Q9: Whether PWTS had published an academic research (One choice)
- Q10: If PWTS had published an academic research, did the research make use of machine learning? (Multiple choice)

### Pragramming skills & experience:

- Q11: Number of years the PWTS have been writing code (One choice)
- Q12: Programming languages that PWTS use on a regular basis? (Multiple choice)
- Q13: IDE that PWTS use on a regular basis? (Multiple choice)
- Q14: Notebook hosting products that PWTS use? (Multiple choice)
- Q15: Data visualization libraries that PWTS use on a regular basis? (Multiple choice)
- Q16: Number of years PWTS have used ML methods? (One choice)
- Q17: ML frameworks that PWTS use on a regular basis? (Multiple choice)
- Q18: ML algorithm that PWTS use on a regular basis? (Multiple choice)
- Q19: Computer vision methods that PWTS use on a regular basis? (Multiple choice)
- Q20: NLP (natural language processing) methods that PWTS use on a regular basis? (Multiple choice)
- Q21: Pre-trained model weights services that PWTS download? (Multiple choice)
- Q22: ML model hubs/repositories that PWTS use often? (Multiple choice)

### Employments:

- Q23: Tittle that most similar to PWTS current role? (Multiple choice)
- Q24: Industry that PWTS currently employed/contracted in? (One choice)
- Q25: Size of company that PWTS employed in? (One choice)
- Q26: The approximate number of individuals responsible for DS workloads at PWTS place of work? (One choice)
- Q27: Whether current employer incorporate ML methods in their business? (One choice)
- Q28: Activities that makes up important part of PWTS at work? (Multiple choice)
- Q29: Current yearly compensation of PWTS? (One choice)

### Other product used (Cloud computing, Data storage...):

- Q30: The approximate number of money that PWTS had spent of ML/cloud computer at home/at work in the past 5 years in USD? (One choice)
- Q31: Cloud computing platforms that PWTS uses? (Multiple choice)
- Q32: Cloud platforms that PWTS had most familiar with/has the best developer experience? (One choice)
- Q33: Cloud computing products that PWTS uses? (Multiple choice)
- Q34: Data storage products that PWTS uses? (Multiple choice)
- Q35: Data products (relational databases, data warehouses, data lakes...) that PWTS are most farmiliar with? (Multiple choice)
- Q36: Business Intelligence tools that PWTS uses? (Multiple choice)
- Q37: Managed machine learning products that PWTS uses? (Multiple choice)
- Q38: Automated Machine Learning tools that PWTS uses? (Multiple choice)
- Q39: Products that PWTS to help with ML models? (Multiple choice)
- Q40: Tools that help PWTS monitor ML models/experiments? (Multiple choice)
- Q41: Responsible/Ethical AI products that PWTS uses in ML practices? (Multiple choice)
- Q42: Specilized hardware that PWTS uses when training ML models? (Multiple choice)
- Q43: Times PWTS used TPU (tensor processing unit)? (One choice)

### Source of information:

- Q44: Favorite DS media sources of PWTS? (Multiple choice)

In total, the survey have 44 questions with number of restrictions related to the choice in some questions. For example, if the answer to Q5 answer is that PWTS's a student, questions related to employment would not be asked.

The reason the number of columns in the dataset (296) is not equal to the number of questions in the survey is that each columns represent an option/a choice in every questions.

## Meaning of row

Each row is an response to the Kaggle state of Data Science Survey - 2022.

## Explore dataset

Firts, let's check if there is any duplicated rows.

In [37]:
print("Number of duplicated rows: ", np.count_nonzero(ds_survey_df.duplicated()))

Number of duplicated rows:  0


We see that there's no duplicated rows.

## Clean dataset
Let's revise the purpose of analyzing the dataset:
- People who is working with data
- What's happening with Machine Learning in different industries
- Find out the best ways for new data scientists like us (in the future) to break into the field.

The purpose of the project is to learn more about the state of Data Science. We see that we only need data fields related to Data Science. That's why we should drop columns whose content isn't usable, like "Duration" column. 


In [38]:
# Drop Duration column
ds_survey_df = ds_survey_df.drop('Duration (in seconds)', axis=1)

Observe that the current naming of each columns is very hard to understand. Let's rename the columns in the dataset for easier comprehension.

In [39]:
# First, let's change the column name of columns with only 1 options
ds_survey_df.rename(columns = 
                     {'Q2': 'AGE',
                      'Q3': 'GENDER', 
                      'Q4': 'COUNTRY',
                      'Q5': 'STUDENT',
                      'Q8': 'EDUCATION LEVEL',
                      'Q9': 'PUBLISH PAPER',
                      'Q11': 'PROGRAMING EXPERIECE',
                      'Q16': 'ML EXPERIENCE',
                      'Q24': 'INDUSTRY',
                      'Q25': 'COMPANY SIZS',
                      'Q26': 'IDV RESPONSIBLE FOR DS',
                      'Q29': 'COMPENSATION'}, 
                     inplace=True)

# Next, let's change the name of the columns with multiple choice

# We are going to standardize the choices to make it easier to read
# First, let's 

TypeError: NDFrame.replace() got an unexpected keyword argument 'columns'

Next, because there are a lot of columns in our dataset, let's simplify the dataset by reorganize our current data frame into mutiple smaller dataframe.
We will reorganize our data into 5 dataframe:
- Education
- Programming skills and Experience
- Employements
- Other products used
- Source of information

In [ ]:
education_df = pd.concat([ds_survey_df.iloc[:, :5], ds_survey_df.iloc[:, 6:28]])
prskill_exp_df = pd.concat([ds_survey_df.iloc[:, :5], ds_survey_df.iloc[:, 28:144]])
employ_df = pd.concat([ds_survey_df.iloc[:, :5], ds_survey_df.iloc[:, 144:158]])
other_df = pd.concat([ds_survey_df.iloc[:, :5], ds_survey_df.iloc[:, 158:283]])
media_df = pd.concat([ds_survey_df.iloc[:, :5], ds_survey_df.iloc[:, 283:]])

Q2     Q3                        Q4   Q5
1      30-34    Man                     India   No
2      30-34    Man                   Algeria   No
3      18-21    Man                     Egypt  Yes
4      55-59    Man                    France   No
5      45-49    Man                     India  Yes
...      ...    ...                       ...  ...
23993  22-24    Man  United States of America  Yes
23994  60-69    Man  United States of America  Yes
23995  25-29    Man                    Turkey   No
23996  35-39  Woman                    Israel   No
23997  18-21    Man                     India  Yes

[23996 rows x 4 columns]

In [ ]:
prskill_exp_df

Q2   Q3       Q4   Q5      Q6_1                               Q44_1  \
1      30-34  Man    India   No       NaN                                 NaN   
2      30-34  Man  Algeria   No       NaN                                 NaN   
3      18-21  Man    Egypt  Yes  Coursera                                 NaN   
4      55-59  Man   France   No  Coursera                                 NaN   
5      45-49  Man    India  Yes       NaN                                 NaN   
...      ...  ...      ...  ...       ...                                 ...   
23993    NaN  NaN      NaN  NaN       NaN                                 NaN   
23994    NaN  NaN      NaN  NaN       NaN                                 NaN   
23995    NaN  NaN      NaN  NaN       NaN  Twitter (data science influencers)   
23996    NaN  NaN      NaN  NaN       NaN                                 NaN   
23997    NaN  NaN      NaN  NaN       NaN                                 NaN   

                                                   Q44_2 Q44_3  \
1                                                    NaN   NaN   
2                                                    NaN   NaN   
3                                                    NaN   NaN   
4                                                    NaN   NaN   
5                                                    NaN   NaN   
...                                                  ...   ...   
23993                                                NaN   NaN   
23994  Email newsletters (Data Elixir, O'Reilly Data ...   NaN   
23995                                                NaN   NaN   
23996                                                NaN   NaN   
23997                                                NaN   NaN   

                                 Q44_4 Q44_5  \
1                                  NaN   NaN   
2                                  NaN   NaN   
3                                  NaN   NaN   
4                                  NaN   NaN   
5                                  NaN   NaN   
...                                ...   ...   
23993  Kaggle (notebooks, forums, etc)   NaN   
23994                              NaN   NaN   
23995  Kaggle (notebooks, forums, etc)   NaN   
23996                              NaN   NaN   
23997                              NaN   NaN   

                                                   Q44_6  \
1                                                    NaN   
2                                                    NaN   
3                                                    NaN   
4                                                    NaN   
5                                                    NaN   
...                                                  ...   
23993  YouTube (Kaggle YouTube, Cloud AI Adventures, ...   
23994  YouTube (Kaggle YouTube, Cloud AI Adventures, ...   
23995  YouTube (Kaggle YouTube, Cloud AI Adventures, ...   
23996  YouTube (Kaggle YouTube, Cloud AI Adventures, ...   
23997                                                NaN   

                                                   Q44_7 Q44_8  \
1                                                    NaN   NaN   
2                                                    NaN   NaN   
3                                                    NaN   NaN   
4                                                    NaN   NaN   
5                                                    NaN   NaN   
...                                                  ...   ...   
23993  Podcasts (Chai Time Data Science, O’Reilly Dat...   NaN   
23994                                                NaN   NaN   
23995                                                NaN   NaN   
23996                                                NaN   NaN   
23997                                                NaN   NaN   

                                                   Q44_9 Q44_10 Q44_11 Q44_12  
1                                                    NaN    NaN    NaN   